src/main/java/net/snowflake/client/jdbc/SnowflakeConnectString.java

```diff
@@ -123,11 +124,25 @@ public static SnowflakeConnectString parse(String url, Properties info) {
         // Account names should not be altered. Set it to a value without org name
         // if it's a global url
         parameters.put("ACCOUNT", account);
-        if (account.contains("_")) {
-          // Update the Host URL to remove underscores if there are any
-          String account_wo_uscores = account.replaceAll("_", "-");
-          host = host.replaceFirst(account, account_wo_uscores);
-        }
+      }
+
+      if (account.contains("_")
+          && parameters.containsKey(
+              SFSessionProperty.ALLOW_UNDERSCORES_IN_HOST.getPropertyKey().toUpperCase())
+          && "false"
+              .equalsIgnoreCase(
+                  (String)
+                      parameters.get(
+                          SFSessionProperty.ALLOW_UNDERSCORES_IN_HOST
+                              .getPropertyKey()
+                              .toUpperCase()))
```

Comment:
nit: since we use `equalsIgnoreCase`, do we still need `toUpperCase`?

Reply:
Also before the change, we call this logic in the block `if (parameters.get("ACCOUNT") == null && account == null && host.indexOf(".") > 0)`. Now we do it outside the if block. Would this cause some problems?

Reply:
@sfc-gh-wshangguan the `toUpper` is for getting the property name. For your other comment of doing this outside the block, it is intentional. We want to alter the hostname irrespective of whether the ACCOUNT name is null or not.